File: haskell/src/Functor.hs
{-# LANGUAGE NoImplicitPrelude #-}

module Functor where

import SimplePrelude

class Functor f where
  fmap :: (a -> b) -> f a -> f b

  (<$>) :: (a -> b) -> f a -> f b
  (<$>) = fmap

  (<$) :: a -> f b -> f a
  (<$) = fmap . const

class Functor f => Applicative f where
  pure :: a -> f a

  ap    :: f (a -> b) -> f a -> f b
  (<*>) :: f (a -> b) -> f a -> f b
  (<*>) = ap

  (*>) :: f a -> f b -> f b
  fa *> fb = (id <$ fa) <*> fb

  (<*) :: f a -> f b -> f a
  (<*) = liftA2 const

class Applicative m => Monad m where
  return :: a -> m a
  return = pure

  (>>=) :: m a -> (a -> m b) -> m b

  (>>) :: m a -> m b -> m b
  ma >> mb = ma >>= (\_ -> mb)

liftA2 :: Applicative f => (a -> b -> c) -> f a -> f b -> f c
liftA2 f fa fb = f <$> fa <*> fb

File: scala/src/main/scala-2.12/pg/tcs/Functor.scala
package tcs

import simulacrum._
import tc.SimplePrelude._

@typeclass
trait Functor[F[_]] {
  def map[A,B](fa: F[A])(f: A => B): F[B]

  final def as[A,B](fb: F[B])(a: A): F[A] =
    map(fb)(_ => a)
}

@typeclass
trait Applicative[F[_]] extends Functor[F] {
  def pure[A](a: A): F[A]

  @op("<*>") def ap[A, B](fa: F[A])(f: F[A => B]): F[B]

  final def *>[A, B](fa: F[A])(fb: F[B]): F[B] =
    apply2(fa, fb)((_,b) => b)

  final def <*[A, B](fa: F[A])(fb: F[B]): F[A] =
    apply2(fa, fb)((a,_) => a)

  final def apply2[A, B, C](fa: => F[A], fb: => F[B])(f: (A, B) => C): F[C] =
    ap(fb)(map(fa)(f.curried))

  final def ap2[A,B,C](fa: F[A])(fb: F[B])(f: F[A => B => C]): F[C] =
    ap(fb)(ap(fa)(f))
}

@typeclass
trait Monad[F[_]] extends Applicative[F] {
  @op(">>=")
  def flatMap[A,B](fa: F[A])(f: A => F[B]): F[B]

  def >>[A,B](fa: F[A])(fb: F[B]): F[B] =
    flatMap(fa)(_ => fb)

  def ret[A](a: A): F[A] = pure(a)
}
